go/ephemeral/user_ek_box_storage.go

```diff
@@ -269,3 +273,31 @@ func (s *UserEKBoxStorage) MaxGeneration(ctx context.Context) (maxGeneration key
 	}
 	return maxGeneration, nil
 }
+
+func (s *UserEKBoxStorage) DeleteExpired(ctx context.Context) (expired []keybase1.EkGeneration, err error) {
+	defer s.G().CTrace(ctx, "DeviceEKStorage#DeleteExpired", func() error { return err })()
+	s.Lock()
+	defer s.Unlock()
+
+	cache, err := s.getCache(ctx)
+	if err != nil {
+		return expired, err
+	}
+
+	keyMap := make(keyExpiryMap)
+	for generation, userEKBoxed := range cache {
+		keyMap[generation] = userEKBoxed.Metadata.Ctime
+	}
+
+	latestMerkleRoot, err := s.G().GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
+	if err != nil {
+		return expired, err
+	}
+	expired = getExpiredGenerations(keyMap, keybase1.Time(latestMerkleRoot.Ctime()))
+	epick := libkb.FirstErrorPicker{}
+	for _, generation := range expired {
+		epick.Push(s.delete(ctx, generation))
```

Comment:
It seems like each of these deletions would rewrite the entire list to the KV store, in a kind of quadratic way? I'm not really worried about the quadratic part since we have N ~= 7, but if it was easy to expose a delete_multi interface, it might be natural to call it here? (That would also save you the trouble of coalescing errors, since there would only be one write. So it might be simpler in the end? Not worth spending time on if it's more complicated.)

The only performance problem I'm really thinking we'll run into here, is that disk writes are sometimes inordinately expensive on crappy Android phones. It could end up being the kind of thing where minimizing round trips is important. But then again if it's all happening in the background, maybe none of it matters at all :)

Reply:
great point and easy fix :)